src/map/script.cpp

```diff
@@ -23728,6 +23730,148 @@ BUILDIN_FUNC(getequiptradability) {
 	return SCRIPT_CMD_SUCCESS;
 }
 
+/**
+ * Swaps position of two array elements.
+ * This function does not guarantee that the array is accessible, so the burden fall to the caller.
+ * @param st Current script state
+ * @param sd The player owning the variable, can be nullptr
+ * @param array_ The array containing elements to be swapped
+ * @param idx1 index of first element
+ * @param idx2 index of second element 
+ */
+void swap_array_element(script_state *st, map_session_data *sd, script_data* array_, int idx1, int idx2) {
+	void *v1, *v2;
+	auto id = reference_getid(array_);
+	auto name = reference_getname(array_);
+	auto ref = reference_getref(array_);
+
+	v1 = get_val2(st, reference_uid(id, idx1), ref);
+	v2 = get_val2(st, reference_uid(id, idx2), ref);
+	set_reg(st, sd, reference_uid(id, idx2), name, v1, ref);
+	set_reg(st, sd, reference_uid(id, idx1), name, v2, ref);
+	script_removetop(st, -2, 0);
+}
+
+/**
+ * Partition an array to be used in quick sort with Hoare's partitioning
+ * This function does not guarantee that the array is accessible, so the burden fall to the caller.
+ * @param st Current script state
+ * @param sd The player owning the variable, can be nullptr
+ * @param array_ The array to be partitioned
+ * @param start Starting index
+ * @param end Ending index
+ */
+int partition_array(script_state *st, map_session_data *sd, script_data *array_, int start, int end) {
+	auto ref = reference_getref(array_);
+	auto id = reference_getid(array_);
+	auto name = reference_getname(array_);
+	int i = start, j = end;
+	int pivotidx = rnd() % (end - start);
+	void* pivot = get_val2(st, reference_uid(id, start + pivotidx), ref);
+	bool is_string = is_string_variable(name);
```

Comment:
Maybe use a  function pointer instead? One that compares two void pointers and return -1,0,1?
Maybe even a delegate would be enough?